src/hotp/hotp.c

```diff
@@ -493,11 +493,13 @@ uint8_t* page = (uint8_t *) current_slot_address;
 
     memcpy (page_buffer, page, SLOT_PAGE_SIZE);
 
-    // check if the secret from the tool is empty and if it is use the old
-    // secret
-uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
-
-    if (secret[0] == 0)
+    // Check if the secret from the tool is empty and if it is use the old secret
+    // Secret could begin with 0x00, so checking the whole secret before keeping the old one in mandatory
+    // The while last loop points secret[-1]. Well, it's right, == result does not matter and mem exits, we need a -1
+    uint8_t* secret = (uint8_t *) (data + SECRET_OFFSET);
+    int8_t i = 20;
+    while (i >= 0 && secret[--i] == 0x00);
```

Comment:
Note: protocol change. Client apps have to send 20 null bytes instead of 1 null byte (empty string) while editing OTP slot. To synchronize with `Nitrokey-app`. `Libnitrokey` already clears whole command buffers thus sending empty string should result in 20*'0'.